### PR TITLE
Add annotations to Sheet and Helper parser; Add errors to Sheet parser

### PR DESCRIPTION
**Pull request by NduatiK**
1. Added annotations to the sheet parser to identify the source file, module and line of a piece of AST
2. Added compile errors and tests for the sheet sigil. You can test it out by modifying the non-rule parts of `MockSheet` and trying to compile.

For example, the following should fail to compile:
```elixir
  ~SHEET"""
  "color-hex-" <> 31 do
    rule-31
    rule-22
  end

  "color-yellow" do
    rule-21
  end

  "named-argument" do
    rule-ime
  end
  """
```

### CODE REVIEW

**Comment by NduatiK** [comment]
@bcardarella, could you check this out?

**Comment by bcardarella** [comment]
oh sorry, I missed this PR

**Comment by NduatiK** [comment]
I just changed it from a draft 30 minutes ago.

File: lib/live_view_native/stylesheet/rules_parser.ex [hunk 1]
```diff
@@ -30,14 +30,18 @@ defmodule LiveViewNative.Stylesheet.RulesParser do
         |> parser.parse()
         |> List.wrap()
         |> Enum.map(&escape(&1))
-      {:error, message} -> raise message
+
```
Comment (bcardarella) [inline lib/live_view_native/stylesheet/rules_parser.ex:33]: can these auto-formatting changes be removed? They make it difficult to do code review

File: lib/live_view_native/stylesheet/sheet_parser/parser.ex [hunk 2]
```diff
@@ -0,0 +1,140 @@
+defmodule LiveViewNative.Stylesheet.SheetParser.Parser do
+  import NimbleParsec
+  alias __MODULE__
```
Comment (bcardarella) [inline lib/live_view_native/stylesheet/sheet_parser/parser.ex:3]: why alias itself?

File: lib/live_view_native/stylesheet/sheet_parser/parser.ex [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  def frozen(combinator \\ empty(), combinator_2) do
+    combinator
+    |> concat(combinator_2)
+    |> pre_traverse({Context, :freeze_context, []})
+    |> post_traverse({Context, :unfreeze_context, []})
+  end
+
+  def put_error(error_parser, error_message, opts \\ []) do
+    error_parser
+    |> post_traverse({Parser.Error, :put_error, [error_message, opts]})
```
Comment (bcardarella) [inline lib/live_view_native/stylesheet/sheet_parser/parser.ex:83]: this should be a single line

File: lib/live_view_native/stylesheet/sheet_parser/parser.ex (hunk 2 above)
Comment (NduatiK) [reply lib/live_view_native/stylesheet/sheet_parser/parser.ex:3]: `post_traverse` expects a mfa tuple and I'd prefer writing `post_traverse({Parser, function, args])` to  `post_traverse({LiveViewNative.Stylesheet.SheetParser.Parser, function, args])`

File: lib/live_view_native/stylesheet/rules_parser.ex (hunk 1 above)
Comment (NduatiK) [reply lib/live_view_native/stylesheet/rules_parser.ex:33]: Done

File: lib/live_view_native/stylesheet/sheet_parser/parser.ex (hunk 2 above)
Comment (bcardarella) [reply lib/live_view_native/stylesheet/sheet_parser/parser.ex:3]: you should use `__MODULE__`

```elixir
post_traverse({__MODULE__, function, args])
```

File: lib/live_view_native/stylesheet/sheet_parser/parser.ex (hunk 3 above)
Comment (NduatiK) [reply lib/live_view_native/stylesheet/sheet_parser/parser.ex]: Good catch. Fixed.

File: lib/live_view_native/stylesheet/rules_parser.ex [hunk 4]
```diff
@@ -30,14 +30,18 @@ defmodule LiveViewNative.Stylesheet.RulesParser do
...
+      {:error, message} ->
+        raise message
```
Comment (bcardarella) [inline lib/live_view_native/stylesheet/rules_parser.ex:35]: formatting removal here

File: lib/live_view_native/stylesheet/rules_parser.ex [hunk 5]
```diff
@@ -30,14 +30,18 @@ defmodule LiveViewNative.Stylesheet.RulesParser do
...
     end
   end
 
   defp escape({operator, meta, arguments}) when operator in [:<>] do
     {operator, meta, Enum.map(arguments, &escape(&1))}
   end
+
   defp escape({Elixir, _meta, expr}), do: expr
+
```
Comment (bcardarella) [inline lib/live_view_native/stylesheet/rules_parser.ex:44]: unnecessary whitespace added from auto-formatting

**Comment by bcardarella** [comment]
We will want to support compilations with and without annotatinos. The annotations are good for dev mode but add unnecessary bytes in production.

I'm thinking that `LiveViewNative.Stylesheet.compile_ast/2` should be changed so the 2nd argument is `[target: :all, annotations: false]` by default.

**Comment by bcardarella** [comment]
are the annotations for the rules parser going to be inherited by the platform parser implementaitons (i.e. swiftui) or do annotaitons have to be reimplemented in each client?

**Comment by NduatiK** [comment]
> The annotations are good for dev mode but add unnecessary bytes in production.

Annotations are only added when outside production ([see here](https://github.com/liveview-native/live_view_native_stylesheet/pull/16/files#diff-89eefee5873942b15ba3e9872b883a857c817d8e22320227862be79aa205959bR9-R17))

File: lib/live_view_native/stylesheet/sheet_parser/parser/annotations.ex [hunk 6]
```diff
@@ -0,0 +1,18 @@
+defmodule LiveViewNative.Stylesheet.SheetParser.Parser.Annotations do
+  alias LiveViewNative.Stylesheet.SheetParser.Parser.Context
+  # Helpers
+
+  def context_to_annotation(%{context: %Context{} = context}, line) do
+    context_to_annotation(context, line)
+  end
+
+  if Mix.env() != :prod do
```
Comment (bcardarella) [inline lib/live_view_native/stylesheet/sheet_parser/parser/annotations.ex:9]: let's not do this approach, referencing the specific environment that annotations will only be available is limiting. Instead if it is an option passed into to `compile_ast` it will allow the developer more control. For example, there may be more environments that annotations want to be stripped from than just `prod`

File: lib/live_view_native/stylesheet/sheet_parser/parser/context.ex [hunk 7]
```diff
@@ -0,0 +1,57 @@
+defmodule LiveViewNative.Stylesheet.SheetParser.Parser.Context do
+  alias __MODULE__
```
Comment (bcardarella) [inline lib/live_view_native/stylesheet/sheet_parser/parser/context.ex:2]: please remove this alias and rely on `__MODULE__` within the code

**Comment by NduatiK** [comment]
> do annotaitons have to be reimplemented in each client

They will be reimplemented for SwiftUI specific ASTs, like `to_ime`, but will be shared for things like variables.
The implementation isn't very DRY at the moment, but I plan to refactor once I have an idea of what Jetpack Compose's stylesheets look like. 

In any case, I plan to rewrite the parser to drop NimbleParsec soon (the current approach works but requires a lot of fiddling to catch the right error). If I'll have to rewrite, then it's not too important to avoid duplication now.

File: lib/live_view_native/stylesheet/sheet_parser/parser/annotations.ex (hunk 6 above)
Comment (NduatiK) [reply lib/live_view_native/stylesheet/sheet_parser/parser/annotations.ex:9]: Makes sense.

`compile_ast` is called after the sheet is parsed and the AST is produced which is too late. What about using application env vars?

Comment (NduatiK) [reply lib/live_view_native/stylesheet/sheet_parser/parser/annotations.ex:9]: We can default to no annotations for `prod` and allow the user to override this if they want.

File: lib/live_view_native/stylesheet/sheet_parser/parser/context.ex (hunk 7 above)
Comment (NduatiK) [reply lib/live_view_native/stylesheet/sheet_parser/parser/context.ex]: Sorry, I missed that one and the one in `Parser.Error`. Removing.

File: lib/live_view_native/stylesheet/sheet_parser/parser/annotations.ex (hunk 6 above)
Comment (bcardarella) [reply lib/live_view_native/stylesheet/sheet_parser/parser/annotations.ex:9]: `compile_ast` is the primary public function that devs will be using so it's where the render state should be sent in. The `annotations` boolean option should be carried down to where it will be used to render/not render annotations

Comment (NduatiK) [reply lib/live_view_native/stylesheet/sheet_parser/parser/annotations.ex:9]: Yes, but to support this, we would have to loop over the AST and strip out the annotations. We can do it, but it won't be pretty.

---
An alternative is to offer it as an option when defining the stylesheet:
```elixir
defmodule MockSheet do
  use LiveViewNative.Stylesheet, :swiftui, annotations: true

  ~SHEET"""
  ...
  """

  def class(_other, _), do: {:unmatched, ""}
end
```

Comment (NduatiK) [reply lib/live_view_native/stylesheet/sheet_parser/parser/annotations.ex:9]: Which do you prefer?

Comment (bcardarella) [reply lib/live_view_native/stylesheet/sheet_parser/parser/annotations.ex:9]: wait, sorry you're correct that the order of operations isn't what I assumed. Let me think about this. Your suggestions of an application var be good. Something like this:

```elixir
Application.get_env(:live_view_native_stylesheet, :annotations, true)
```

This way it is always defaulted to rendering annotations then you opt-in for prod-like environments. I'm thinking we may want something like this for pretty-print too which can be outside the PR. So the default rendering will always be a annotated pretty-print but then you can shrink the asset with certain config options

Comment (bcardarella) [reply lib/live_view_native/stylesheet/sheet_parser/parser/annotations.ex:9]: @NduatiK let's go with the config var, it's a single choke point. The way we're looking at stylehseets is probably having multiple sheets, at least one per LV itself. So having a single place to set the render type for environments is better

Comment (NduatiK) [reply lib/live_view_native/stylesheet/sheet_parser/parser/annotations.ex]: 👍 Done

Comment (NduatiK) [reply lib/live_view_native/stylesheet/sheet_parser/parser/annotations.ex]: All done

**Comment by bcardarella** [comment]
This is looking good, I think we just need a single assertion for the rendering without annotations:

```elixir
Application.put_env(:live_view_native_stylesheet, :annotations, false)
# assertion without annotations
Application.delete_env(:live_view_native_stylesheet, :annotations)
```

**Comment by NduatiK** [comment]
> just need a single assertion for the rendering without annotations

Yup, already [added](https://github.com/liveview-native/live_view_native_stylesheet/pull/16/files#diff-c0a8f7abcd8f3fbfa4fa0b1d4b788f6695166f0e6a73806d7e1f97a38227bf42R35-R37)

**Review by bcardarella — APPROVED** [review]
